docker-compose-rule-core/src/main/java/com/palantir/docker/compose/execution/DockerCommandLocations.java

```diff
@@ -1,47 +1,75 @@
- * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
  */
-package com.palantir.docker.compose.execution;
 
-import static java.util.Arrays.asList;
+package com.palantir.docker.compose.execution;
 
 import java.io.File;
-import java.util.List;
-import java.util.Optional;
-import java.util.function.Predicate;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.immutables.value.Value;
 
-public class DockerCommandLocations {
-    private static final Predicate<String> IS_NOT_NULL = path -> path != null;
-    private static final Predicate<String> FILE_EXISTS = path -> new File(path).exists();
+@Value.Immutable
+public abstract class DockerCommandLocations {
 
-    private final List<String> possiblePaths;
+    private static final Pattern PATH_SPLITTER = Pattern.compile(File.pathSeparator);
 
-    public DockerCommandLocations(String... possiblePaths) {
-        this.possiblePaths = asList(possiblePaths);
+    @Value.Default
+    protected Map<String, String> env() {
+        return System.getenv();
     }
 
-    public Optional<String> preferredLocation() {
+    @Value.Derived
+    protected String path() {
+        String path = env().get("PATH");
+        if (path == null) {
+            path = env().get("path");
+        }
+        if (path == null) {
+            throw new IllegalStateException("No path environment variable found");
+        }
+        return path;
+    }
 
-        return possiblePaths.stream()
-                .filter(IS_NOT_NULL)
-                .filter(FILE_EXISTS)
-                .findFirst();
+    @Value.Check
+    protected void pathIsNotEmpty() {
+        if (path().isEmpty()) {
+            throw new IllegalStateException("Path variable was empty");
+        }
     }
 
-    @Override
-    public String toString() {
-        return "DockerCommandLocations{possiblePaths=" + possiblePaths + "}";
+    @Nullable
+    protected abstract String locationOverride();
+
+    @Value.Default
+    protected Stream<String> macSearchLocations() {
+        return Stream.of("/usr/local/bin", "/usr/bin");
     }
+
+    private Stream<String> pathLocations() {
+        Stream<String> pathLocations = Stream.concat(PATH_SPLITTER.splitAsStream(path()), macSearchLocations());
+        if (locationOverride() == null) {
+            return pathLocations;
+        }
+        return Stream.concat(Stream.of(locationOverride()), pathLocations);
+    }
+
+    public Stream<Path> forCommand() {
```

Comment:
Purely because it was extracted out the middle of a stream, I'll make it a list.